platform/linux-generic/odp_ipsec.c

```diff
@@ -565,18 +593,40 @@ static int ipsec_in_esp(odp_packet_t *pkt,
 	param->auth_iv_ptr = state->iv;
 
 	state->esp.aad.spi = esp.spi;
-	state->esp.aad.seq_no = esp.seq_no;
 	state->in.seq_no = odp_be_to_cpu_32(esp.seq_no);
+	if (ipsec_sa->esn)
+		ipsec_sa->in.seq64 = ipsec_compute_esn(ipsec_sa, state->in.seq_no);
+
+	/* AAD with ESN */
+	if (ipsec_sa->esn && (!ipsec_sa->seqh_len))
+		state->esp.aad.seq_no64 = odp_cpu_to_be_64(ipsec_sa->in.seq64);
+	else if (!ipsec_sa->esn)
+		state->esp.aad.seq_no = esp.seq_no;
 
 	param->aad_ptr = (uint8_t *)&state->esp.aad;
 
+	/* Inbound ICV computation includes ESN higher 32 bits for individual algo's */
```

Comment:
This is true also for combined mode algorithms (it is just done in a different way) so the comment may be a bit misleading.

Reply:
ack, will correct this comment.

Reply:
Now the comment says:
+       /* Inbound ICV computation includes ESN higher 32 bits. */

It is a true statement, but it should be better related to the subsequent code which is for non-combined mode algorithms only. How about this: "Insert high-order bits of ESN before the ICV for ICV check with non-combined mode algorithms."

Reply:
ack, will change it.